### PR TITLE
Add MCAP v0 local file support

### PR DESCRIPTION
**Pull request by jtbandes**
- Refactor commonly reused channel/schema parsing and decompression functions into `parseChannel` into new `@foxglove/mcap-support` package
- Fix check for missing schemas to support schema id 0
- Add `Mcap0StreamedDataProvider` & `Mcap0IndexedDataProvider`
- Add `McapDataProvider` which detects the mcap version and delegates to `Macp0StreamedDataProvider`, `Mcap0IndexedDataProvider `, or `McapPre0DataProvider`

### CODE REVIEW

File: packages/mcap-support/src/parseChannel.ts [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    if (index === 0) {
+      datatypes.set(rootName, { name: rootName, definitions });
+    } else if (name != undefined) {
+      datatypes.set(name, { name, definitions });
+    }
+  });
+  return datatypes;
+}
+
+export function parseChannel(channel: Channel): ParsedChannel {
```
Comment (defunctzombie) [inline packages/mcap-support/src/parseChannel.ts:42]: Comment? What is parseChannel? I'd have thought I already parsed a channel when I got it from the mcap reader library.

File: packages/studio-base/src/players/FoxgloveWebSocketPlayer/index.ts [hunk 2]
```diff
@@ -120,9 +58,9 @@ export default class FoxgloveWebSocketPlayer implements Player {
 
   private _unresolvedSubscriptions = new Set<string>();
   private _resolvedSubscriptionsByTopic = new Map<string, SubscriptionId>();
-  private _resolvedSubscriptionsById = new Map<SubscriptionId, ParsedChannel>();
-  private _channelsByTopic = new Map<string, ParsedChannel>();
-  private _channelsById = new Map<ChannelId, ParsedChannel>();
+  private _resolvedSubscriptionsById = new Map<SubscriptionId, ResolvedChannel>();
```
Comment (defunctzombie) [inline packages/studio-base/src/players/FoxgloveWebSocketPlayer/index.ts:61]: `HydratedChannel` :D

Comment (jtbandes) [reply packages/studio-base/src/players/FoxgloveWebSocketPlayer/index.ts:61]: hydrated with our tears

File: packages/studio-base/src/randomAccessDataProviders/McapDataProvider.ts [hunk 3]
```diff
@@ -2,214 +2,123 @@
...
 } from "@foxglove/studio-base/randomAccessDataProviders/types";
-import { RosDatatypes } from "@foxglove/studio-base/types/RosDatatypes";
+
+import Mcap0IndexedDataProvider from "./Mcap0IndexedDataProvider";
+import Mcap0StreamedDataProvider from "./Mcap0StreamedDataProvider";
+import McapPre0DataProvider from "./McapPre0DataProvider";
 
 type Options = { file: File };
 
+class FileReadable {
```
Comment (defunctzombie) [inline packages/studio-base/src/randomAccessDataProviders/McapDataProvider.ts:22]: Separate file for organization?

**Review by defunctzombie — APPROVED** [review]
Some more comments around the `mcap-support` functions might be nice or generally planning out the structure of that module.

File: packages/studio-base/src/randomAccessDataProviders/McapDataProvider.ts (hunk 3 above)
Comment (jtbandes) [reply packages/studio-base/src/randomAccessDataProviders/McapDataProvider.ts:22]: 🤷 when something is a few lines and (for now) only relevant to one class, I don't feel bad about keeping it in the same file

**Pull request merged**